en/02_Developer_Guides/02_Controllers/07_CMS_JSON_APIs.md

```diff
@@ -0,0 +1,154 @@
+        // Note: would normally validate that $itemID is a valid integer and that $obj exists
+        $obj = MyDataObject::get()->byID($itemID);
+        $data = ['ID' => $obj->ID, 'Title' => $obj->Title];
+        return $this->jsonSuccess(200, $data);
+    }
+}
+```
+
+Remember to add any public methods that are used as endpoints to `private static array $allowed_actions`;
+
+## Permission checks
+
+Incorporate essential permission checks, such as `canEdit()`, into all relevant endpoints to ensure secure access control.
+
+When returning DataObjects as JSON, remember to invoke `canView()` on each DataObject. In a CMS context where the number of DataObjects is typically limited, the performance impact of these checks should not be a significant concern. If the permission check fails then call `$this->jsonError(403);` to return a 403 status code.
+
+## Return values and error handling
+
+All public endpoint methods must be declare a return type of [`HTTPResponse`](api:SilverStripe\Control\HTTPResponse).
+
+All return values must utilise `jsonSuccess()` to create the `HTTPReponse` as this method is used to standarise JSON responses in the CMS.
+
+Do not throw exceptions in the controller, as this leads to a poor content-editor experience. Instead all non-success conditions must call `jsonError()`.
+
+### Using `jsonSuccess()`
+
+When using the optional `$data` parameter in `jsonSuccess()` to return JSON in the response body, do not add any "success metadata" around it, for example `['success' => true, 'data' => $data]`. Instead, solely rely on standard HTTP status codes to clearly indicate the success of the operation.
+
+For scenarios where no JSON data is returned in the response body upon success, use the status code 201 withou the `$data` parameter i.e. `return $this->jsonSuccess(201);`. Alternatively, when the response includes JSON data, usually return a 200 status code i.e. `return $this->jsonSuccess(200, $data);`.
+
+### Using `jsonError()`
+
+The `return` keyword is omitted for `jsonError()` because internally it triggers an exception, subsequently caught and converted into an `HTTPResponse` object.
+
+Generally you should not include a message outlining the nature of the error when calling `jsonError()`, instead just include the appropriate HTTP status code for the error type, for example call `$this->jsonError(403)` if a permission check fails.
+
+If you do include a message, remember that error messages are only intented for developers so do not use the `_t()` function to make them translatable. Do not use any returned messages on the frontend for things like toast notifications, instead thoses messages should be added directly in JavaScript.
+
+[info]
+Despite the slightly convuluted JSON format returned of `jsonError()` with multiple nodes, its usage remains consistent with `FormSchema`. It's better to use this method for uniformity rather than introducing separate methods for `FormSchema` and non-FormSchema failures.
```

Comment:
```suggestion
Despite the slightly convoluted JSON format returned of `jsonError()` with multiple nodes, its usage remains consistent with `FormSchema`. It's better to use this method for uniformity rather than introducing separate methods for `FormSchema` and non-FormSchema failures.
```

Reply:
Updated